### ISSUE TITLE
Macro `for_all_cmp_combinations` is too heavy.

### ISSUE THREAD

**Issue by sundy-li** [skyzh/type-exercise-in-rust#3]
1. It's really a heavy macro
2. It did not take consider unsigned types, such as `UInt32 CMP  Int32`, the cast type may be `Int64`.  (seems DataType did not have unsigned types for now, it's a bit more like MySQL)

**Comment by skyzh** [comment]
> It's really a heavy macro

Exactly. Considering the `build_binary_expression` is only called once when a plan is built, I think it's okay to leave such macro here.

> It did not take consider unsigned types, such as UInt32 CMP Int32, the cast type may be Int64. (seems DataType did not have unsigned types for now, it's a bit more like MySQL)

The `DataType` follows Postgres's data type design, so it doesn't take unsigned data into consideration. There's no unsigned type in Postgres.

To support unsigned data types in this framework, I would suggest the following ways:

* unsigned / signed is a kind of logical type. They should be stored as `i32` no matter it's signed or unsigned in logical type.
* signed / unsigned is only useful in integer types, so we should implement a separate set of `CmpXx` for integer types. The new `CmpIntegerLe` struct should use const generics to determine whether each input is signed or unsigned type.
* after that, we can refine the macro to generate unsigned / signed comparison.

**Comment by skyzh** [comment]
... and if the expression framework is used on OLTP databases like TiKV's coprocessor, I think it's more suitable to only having `I64Array` in `ArrayImpl`. All integers, no matter `i16` or `i32`, should be casted to `i64` before proceeding to execution. In this way, the `for_all_cmp_combinations` could be simplified a lot.

**Comment by sundy-li** [comment]
I get it. 

For datatypes:

In OLAP database, I think it's better to use more strict datatypes.  It might not be an issue in this exercise, let's keep it as simple as it could.

For combinations:
Since it's   `10 primitives * 10 primitives` (if we have unsigned as types),  We had to have some [codegen tools](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/codegen/src/writes/arithmetics_type.rs) to generic the type binding of `cmp/arithmetics::*` .
